src/datastore.rs

```diff
@@ -0,0 +1,609 @@
+//! Storage structures for UFO data and images.
+
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    sync::{Arc, RwLock},
+};
+
+use crate::error::{DataStoreError, ImageStoreError};
+use crate::Error;
+
+/// A generic file store for UFO [data][spec_data] and [images][spec_images],
+/// mapping [`PathBuf`] keys to [`Vec<u8>`] values.
+///
+/// The store provides a basic HashMap-like interface for checking data in and out.
+/// If initialized from disk, data can be loaded eagerly or lazily, as in, on access.
+/// It will remember the root data directory for this purpose. This complicates the
+/// accessor methods somewhat, because 1. access can fail with an IO error and 2.
+/// insertion can fail. Data is wrapped in a [`std::sync::Arc`] to help on-demand loading.
+///
+/// Note that it tracks files, not directories. Data paths you insert must not have
+/// any existing path in the store as an ancestor, or you would nest a file under a
+/// file. Images must always be in a flat directory. The paths are always relative to
+/// a UFO's data directory.
+///
+/// # Example
+///
+/// Consider a UFO on disk with the following structure:
+///
+/// ```text
+/// Test.ufo/
+/// ├── data/
+/// │   ├── a.txt
+/// │   ├── b.bin
+/// │   └── com.testing.random/
+/// │       ├── c.txt
+/// │       └── zzz/
+/// │           └── z.txt
+/// ├── images/
+/// │   ├── image1.png
+/// │   ├── image2.png
+/// │   └── image3.png
+/// ├── glyphs/
+/// │   ├── a.glif
+/// │   └── contents.plist
+/// ├── layercontents.plist
+/// └── metainfo.plist
+/// ```
+///
+/// The `data` subfolder will be represented in a [`DataStore`] like so:
+///
+/// * `PathBuf::from("a.txt")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("b.bin")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("com.testing.random/c.txt")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("com.testing.random/zzz/z.txt")` → `b"<content>".to_vec()`
+///
+/// The `images` subfolder will be represented in an [`ImageStore`] like so:
+///
+/// * `PathBuf::from("image1.png")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("image2.png")` → `b"<content>".to_vec()`
+/// * `PathBuf::from("image3.png")` → `b"<content>".to_vec()`
+///
+/// [spec_data]: https://unifiedfontobject.org/versions/ufo3/data/
+/// [spec_images]: https://unifiedfontobject.org/versions/ufo3/images/
+#[derive(Debug, Clone)]
+pub struct Store<T, E> {
+    items: HashMap<PathBuf, Arc<RwLock<Item<E>>>>,
```

Comment:
We definitely don't want a lock here. If we want interior mutability, we could consider a `RefCell`.

I get wanting to keep `get()` and `iter()` non `&mut self`.

Maybe the easiest way to do this is to split up `Item`, like:

```rust
struct ItemInner {
    Loaded(Arc<[u8]>),
    NotLoaded,
    Error(Arc<Error>),
}

Item {
    inner: RefCell<ItemInner>,
}

impl DataStore {
    fn get(&self, key: &Path) -> Result<Option<Arc<[u8]>>, Error> {
        let item = match self.items.get(key) {
            Some(item) => item,
            None => return Ok(None),
        };

        if let Item::NotLoaded = item {
            match self.impl_type.try_load_item(path) {
                Ok(data) => *item = Item::Loaded(data),
                Err(e) => *item = Item::Error(e),
            }
        }

        match item {
            Item::Loaded(data) => Ok(Some(data.clone())),
            Item::Error(err) => Err(err.clone()),
            Item::NotLoaded => unreachable!(),
        }
    }
```

I think that would work? And be a bit simpler maybe?

Reply:
Will using a RefCell impede multithreading when e.g. multiple threads access the data store while some item us `NotLoaded` still?

Reply:
`Font` is not `Sync`, and cannot be used concurrently from multiple threads.

Reply:
Hello! I was looking at this with Nikolaus the other day and now I'm curious too to know more about RefCell vs RwLock. The parallel scenarios we had in mind are:

* open a UFO with norad in the main thread, then launch a thread pool to build individual glyphs in parallel
* open a designspace with several UFOs attached, and launch thread to interpolate instances from that.

All the threads would share the UFO and also this store, and potentially all would try to read from the store. Does that make sense?

Reply:
(I just saw your reply about the Font not being Sync)

Reply:
so the images/data is definitely something that would complicate that. Everything else in the font object doesn't require any mutation, and so if you used some kind of scoped thread pool you should be able to do things like get glyphs from multiple threads, although I haven't played around with this.

If you need images/data to build glyphs, then we could look at the `RwLock` design, but it isn't anything we've needed anywhere else in the library so far.

In practice, it might be useful to know how the image/data resources are commonly used? are items there commonly shared between many glyphs?

Reply:
To be honest in the workflows I know of, it's not the case at all, usually in the data folder we put XML dumps of fontTools data to be merged into the final TTF, or background images that are useful while designing but not while compiling, so it would probably not be an issue.

Reply:
Other foundries might be doing different things though, not sure about that

Reply:
Using just a RefCell<Item> now. I indeed also had to remove the test where we launched 100 threads and read stuff from an outer Font because RefCell could not be passed over :)

Reply:
Yes sorry about saying we should add that test, I didn't know concurrent access was not in the scope